dvc/utils/serialize/_yaml.py

```diff
@@ -3,6 +3,7 @@
 from contextlib import contextmanager
 
 from funcy import reraise
+from jinja2 import Template
```

Comment:
Hm, not quite sure that jinja template is the best format we can use in dvc files, feels like it is still a bit unusual, at least coming from python. It is mature and powerful though, no doubt about it :slightly_smiling_face:

Reply:
Yeah, I went with this because it was relatively easy to use and mature. We also touched on this in issue #3633. If you/anyone has any recommendations I'd be open to taking a crack at implementing it.

Reply:
Hey @efiop 

Would a  custom syntax be more acceptable? I have been using this fork quite a lot recently, if we were to go down the path of a custom templating syntax I think the following features are important. I welcome any feedback :-)

  - Support for looping flagged command-line arguments. ie: I currently use this pattern frequently. (I'll show it in Jinja for now)
```yaml
vars:
  repeated_arg:
    - thing
    - another_thing
    - another_other_thing
stages:
  foo:
    wdir: .
    cmd: >-
      python foo.py
      {% for x in repeated_arg %}
      -a "{{ x }}"
      {% endfor %}

# Renders as
vars:
  repeated_arg:
    - thing
    - another_thing
    - another_other_thing
stages:
  foo:
    wdir: .
    cmd: >-
      python foo.py
      -a "thing"
      -a "anothet_thing"
      -a "another_othert_thing"
```
  - Similarly, looping over key-value pairs would be handy. For example, if a command has optional flagged args you could template the function call in the `stage` and simply update the `vars` as need.
  - On a similar vein, supporting a combination of positional and flagged args would be handy. Potentially using a list of tuples, ie:

```
vars:
  mixed_args:
    - pos-arg-1
    - pos-arg-2
    - ['--flag-arg-1', 42]
    - ['--flag-arg-2', 3.14159]
```
  - The templated syntax should stand out so it is easy to identify at a glance
  - Templated `dvc.yaml` should still be valid yaml (not sure about this, but intuitively I like the idea and have been striving for it)
  - Be able to pickup the stdout of a stage and refer to it in subsequent stages. I have not done this but I have seen it in Kubeflow pipelines and like the idea. ie:

```yaml
stages:
  one:
    wdir: .
    cmd: "echo Bailey"
  two:
    wdir: .
    cmd: "echo {{ stages.one.stdout }} is the greatest doggo in the universe!!!"

# After stage one completes stage two renders as

  two:
    wdir: .
    cmd: "echo Bailey is the greatest doggo in the universe!!!"
```
  _Though maybe this is getting a bit distracted from the initial goal_

---

Also pulling in interested parties from #3633

@dsuess @dmpetrov @karajan1001 @skshetry  @elgehelge @jcpsantiago

Reply:
Haha, I only had used `jinja2` before, used in an inner annotation platform. It is a powerful tool, maybe over powerful and complex here?